sdks/python/apache_beam/transforms/util.py

```diff
@@ -864,3 +867,225 @@ def add_window_info(element, timestamp=DoFn.TimestampParam,
+    """
+    PTransfrom to find the matches as per regex and return as a KV pair.
+    """
+    def _process(element):
+      match = regex.match(element)
+      if match:
+        yield (match.group(keyGroup), match.group(valueGroup))
+
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def matches(pcoll, regex, group=None):
+    """
+    Returns the matches if the entire line matched the Regex. Returns the
+    entire line (group 0 by default). Group can be integer value and a string
+    value.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+      group: (optional) name of the group, it can be integer or a string value.
+    """
+    regex = Regex._regex_compile(regex)
+    group = group or 0
+
+    def _process(element):
+      m = regex.match(element)
+      if m:
+        yield m.group(group)
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def all_matches(pcoll, regex):
+    """
+    Returns the matches if the entire line matches the Regex. Returns all
+    groups.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+    """
+    regex = Regex._regex_compile(regex)
+    return pcoll | Regex._matches_all_object(regex=regex)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(typehints.KV[K, V])
+  @ptransform_fn
+  def matches_kv(pcoll, regex, keyGroup, valueGroup):
+    """
+    Returns the matches if the entire line matches the Regex. Returns the
+    specified groups as the key and value pair.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+      keyGroup: The Regex group to use as the key. Can be int or str.
+      valueGroup: The Regex group to use the value. Can be int or str.
+    """
+    regex = Regex._regex_compile(regex)
+    return pcoll | Regex._matches_kv_object(regex=regex, keyGroup=keyGroup,
+                                            valueGroup=valueGroup)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def find(pcoll, regex, group=None):
+    """
+    Returns the matches if a portion of the line matches the Regex. Returns
+    the entire line (group 0 by default). Group can be integer value and a
+    string value.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+      group: (optional) name of the group, it can be integer or a string value.
+    """
+    regex = Regex._regex_compile(regex)
+    group = group or 0
+
+    def _process(element):
+      r = regex.search(element)
+      if r:
+        yield r.group(group)
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def find_all(pcoll, regex):
```

Comment:
What's the difference between this and all_matches?

Reply:
`find_all()` should return results if the portion of the line matches the regular expression while `all_matches()` returns if the entire line matches the expression.
Fixed the code by using `re.finditer` in the `find_all()` and `re.match` in the `all_matches()`

Reply:
As as I read this code, for every input element, it returns the concatenated list of all groups of all matches it found, which feels odd. E.g. consider the regex "a(b*)" applied to "abb ax abbb". This will result in a single output ["abb", "bb", "a", "", "abbb", "bbb"].  

On the other hand, the Java implementation would only return ["abb", "bb"] (which isn't clear from its docs, and definitely doesn't match the name FindAll let alone re.findall). 

I think what makes sense here is to provide a findall() that is basically Map(lambda line: re.findall(regex, line)) and finditer that uses finditer, and the latter returns match objects which can be used in a follow-up lambda to extract groups, etc. rather than a flat list of indexed groups that would be more difficult to consume.

Reply:
So what I understand from your feedback is:
1. for the 'Regex.find_all()', we need to change `re.finditer` to `re.findall` and return the result as a list.
2. create a new method `Regex.finditer()` which returns "match objects" instances.. 

The issue with the 'MatchObject' is we can not pickle it.
Showing error `PicklingError: Can't pickle <built-in method match of _sre.SRE_Pattern`
So either we can make a custom match object class and return it or uses KV pairs? Where K as the string of group and V as a tuple containing all groups?
For example:
lets say sting is "abb ax abbb" & regex expression is "a(b*)", so the results would be like:
`[
   KV("abb", ('bb',)), 
   KV("a", ('',)), 
   KV("abbb", ('bbb',)
]`
or maybe we can include group 0 in the values as well.
`[
   KV("abb", ('abb', 'bb',)), 
   KV("a", ('ab', '')), 
   KV("abbb", ('abbb', 'bbb')
]`

Your thoughts?